# DS3000 Day 8

Oct 11, 2024

Admin
- Lab 2 Grades (and Solutions!) Posted. **Please** check the solutions and compare with your own before submitting a regrade request.
- Take-Home Coding Exam due **Sunday, Oct. 13, by 11:59 pm**
- There will be a virtual visitor in Dr. Gerber's Tuesday section (**11:45 am, INV 019**) all students are welcome to come either in-person or attend via [Zoom (Link)](https://northeastern.zoom.us/j/96399994252). The visitor is Alumni Justin Chen, who graduated last year and works for a **major** Baseball team.
- Meet with assigned TA by next **Friday, Oct. 18** (all students *must* attend the meeting)
- Phase II of the project due **Friday, Oct. 25** (Group submission to Gradescope, see ProjectGuidelines on Canvas)
- Homework 3 due **Tuesday, Oct. 29**

Push-Up Tracker
- Section 03: 4
- Section 05: 2

Content:
- Linear Algebra Basics Continued...
  - By Hand and In Python

In [2]:
# packages for today
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

import requests
from bs4 import BeautifulSoup
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.backends.backend_pdf import PdfPages
import seaborn as sns
import plotly.express as px

## Machine Learning Overview

All data needs to be represented mathematically in order to perform Machine Learning. What is *Machine Learning* anyway?

- Start with data: observed measures from the real world, store them as vectors/matrices
- Develop a strategy of analysis that leads to choosing an appropriate algorithm
- "Learn" $\rightarrow$ improve the algorithm, identify patterns in the data by performing mathematical operations on the vectors/matrices
- Make predictions on new observations based on the results of the model, or clarify understanding of patterns

### Example ML Algorithms

Here are some example ML algorithms, some of which we'll be studying. They all rely on the mathematical concepts we'll be covering over the next several classes (as you'll see).

#### Supervised Learning (Prediction)
- Linear*, Multiple*, Polynomial* Regression (Regression: predicting 1 numeric feature with 1 or more numeric/categorical features)
- Linear Perceptron* (Classification: predicting 1 categorical feature with 2 or more numeric/categorical features)
- Gradient Descent* and Neural Networks (Feed-Forward*) (Regression or Classification)
- Logistic Regression**
- Nearest Neighbor Classifiers**
- Decision Trees, Random Forests**
- Support Vector Machines
- Bayesian Models

#### Unsupervised Learning (Inference)
- Principal Component Analysis* (Dimensionality Reduction: Understanding how numeric features relate to each other)
- K-means, Clustering**
- Factor Analysis
- Gaussian Mixture Models
- Kernel Estimation

#### Key
- "*" means we will be (or hope/plan to) cover it in this course
- "**" means we'll have talked about the basic mathematical concepts that make these methods up, and are simple enough that I believe you could learn it on your own outside of class (and potentially apply it in a project setting)
- The rest are not too terribly difficult, though perhaps would require a bit more effort on your part to figure out.

# Last Time: Linear Algebra Basics
## Vector Geometry (on board and in python)

Previously, we looked at representing data as vectors, and already talked about calculating distances between vectors such as  $x = \begin{bmatrix} x_1 \\ x_2 \end{bmatrix}$ and $x' = \begin{bmatrix} x_1' \\ x_2' \end{bmatrix}$. Data are usually represented as vectors (or matrices). For example, from the penguin data:

$x_0 = \begin{bmatrix} -0.895341 \\ 0.778650 \\ -1.428247 \\ -0.567127 \end{bmatrix}$

Which means that the first penguin had above average bill depth, but below average everything else. In python, we usually store vectors using NumPy arrays (let's round to three decimals to make it look cleaner):


In [10]:
df_penguin = sns.load_dataset('penguins')
df_penguin.dropna(axis=0, inplace=True)
# only focus on numerical features (for now)
col_num_list = ['bill_length_mm', 'bill_depth_mm', 'flipper_length_mm', 'body_mass_g']
df_penguin_num = df_penguin.loc[:, col_num_list]
# standardize
# subtracting the mean and dividing each feature by the standard deviation (standardization)
df_penguin_num_scaled = pd.DataFrame()
for feat in df_penguin_num.columns:
    df_penguin_num_scaled[f'{feat}_scaled'] = (df_penguin_num[feat] - df_penguin_num[feat].mean()) / df_penguin_num[feat].std()

df_penguin_num_scaled.head()

bill_length_mm_scaled  bill_depth_mm_scaled  flipper_length_mm_scaled  \
0              -0.894695              0.779559                 -1.424608   
1              -0.821552              0.119404                 -1.067867   
2              -0.675264              0.424091                 -0.425733   
4              -1.333559              1.084246                 -0.568429   
5              -0.858123              1.744400                 -0.782474   

   body_mass_g_scaled  
0           -0.567621  
1           -0.505525  
2           -1.188572  
4           -0.940192  
5           -0.691811

In [11]:
vec_penguin0 = np.array(df_penguin_num_scaled.iloc[0, :]).round(3)
vec_penguin0

array([-0.895,  0.78 , -1.425, -0.568])

### Notation

Generally, we use **lowercase** letters to represent vectors, (for example, instead of writing out "Penguin 0", we would call the Austria vector $\vec{x}_0$, and **uppercase** letters to represent matrices, such as when we consider the data set containing the first two penguins (standardized):

$X = \begin{bmatrix} -0.895 & 0.779 & -1.428 & -0.567 \\ -0.822 & 0.119 & -1.068 & -0.506 \end{bmatrix}$

Note that by convention also, vectors are **column vectors**, but that when we combine vectors into a data matrix, the vectors are included as the **rows**.

## Vector Operations (in short)
### Vector Addition

In order to add two vectors together, they must be the same **dimension**. For example, if each vector $\vec{x}_i \in \mathbb{R}^2$):

$\vec{x}_1 = \begin{bmatrix} 3 \\ 4 \end{bmatrix}$
$\vec{x}_2 = \begin{bmatrix} 7 \\ 1 \end{bmatrix}$

$\vec{x}_1 + \vec{x}_2 = \begin{bmatrix} 10 \\ 5 \end{bmatrix}$

In [4]:
x1 = np.array([3, 4])
x2 = np.array([7, 1])
x1 + x2

array([10,  5])

### Vector "Multiplication"

You can easily multiply scalars (real numbers) to vectors:

$c\vec{x}_1 = \begin{bmatrix} 3(c) \\ 4(c) \end{bmatrix}$

In [5]:
# scalar multiplication (what would 3 of penguin 1 look like?)
c=3
c*x1

array([ 9, 12])

But when we talk about "multiplying" vectors together, there may be more than one interpretation of that. When we think of multipying matrices elementwise, this is called the **Hadamard product**:

$\vec{x}_1 = \begin{bmatrix} 3 \\ 4 \end{bmatrix}$
$\vec{x}_2 = \begin{bmatrix} 7 \\ 1 \end{bmatrix}$

$\vec{x}_1 \odot \vec{x}_2 = \begin{bmatrix} 21 \\ 4 \end{bmatrix}$

In [6]:
x1 * x2

array([21,  4])

A more common (and useful) operation is the **dot product**. For two vectors $x$ and $y$, the dot product is:

$x\cdot y = \sum_i x_i \times y_i$

That is, the sum of all the pairwise products of the vectors (or, equivalently, the sum of the Hadamard product vector). Let's use a simple, two-dimensional example instead of our countries for a moment:

$x = \begin{bmatrix} 3 \\ 4 \end{bmatrix}$

$y = \begin{bmatrix} 7 \\ 1 \end{bmatrix}$

$x \cdot y = 3\times7 + 4\times1 = 21 + 4 = 25$

This is a common operation in ML that we'll use quite a bit!

In [7]:
x = np.array([3, 4])
y = np.array([7, 1])
np.dot(x, y)

25

Another useful operation we might use on occasion is finding the length of a vector:

$||x|| = \sqrt{\sum_i x_i^2}$

For example, if:

$x = \begin{bmatrix} 3 \\ 4 \end{bmatrix}$
$||x|| = \sqrt{3^2  + 4^2} = \sqrt{9 + 16} = \sqrt{25} = 5$

We can also use graphing to understand this a little bit better (professor will show this on the board).

In [8]:
#in numpy
np.linalg.norm(x)

5.0

This is also called finding the $\ell 2$-**norm** of the vector. There are other **norms** which we may talk about if necessary as we proceed.

## Data as Matrices

We've already talked at length about representing data as vectors, for example:

$x_0 = \begin{bmatrix} -0.895341 \\ 0.778650 \\ -1.428247 \\ -0.567127 \end{bmatrix}$

Is the first penguin in the Seaborn data set. While some machine learning algorithms can deal with observations/vectors individually (on an iterative basis), many machine learning algorithms consider the observations collectively (as a matrix). This is why general practice is to treat data as **matrices**:

$$X = \begin{bmatrix} -0.895 & 0.779 & -1.428 & -0.567 \\ -0.822 & 0.119 & -1.068 & -0.506 \\ \vdots & \vdots & \vdots & \vdots & \vdots \end{bmatrix}$$

where each row of the matrix is an observation, and each column a feature. Many of the common operations that make up machine learning algorithms require treating data as matrices. It is important also to remember that, in fact, *vectors **are** matrices*, they are simply matrices with one of the dimensionalities equal to one. For example:

$$\vec{a} = \begin{bmatrix} a_1 \\ a_2 \\ a_3 \end{bmatrix}$$

Is both a 3-dimensional vector and a $3 \times 1$ matrix. It's transpose (flipping the rows and columns): $\vec{a}^T = \begin{bmatrix} a_1 & a_2 & a_3 \end{bmatrix}$ is both a 3-dimensional vector and a $1 \times 3$ matrix. The below:

$$A = \begin{bmatrix} 1 & 2 & 3 \\ -4 & -5 & -6 \end{bmatrix}$$
$$B = \begin{bmatrix} -5 & 6 & 1 & -2 \\ 0 & 1 & 1 & 0 \\ 8 & 6 & 4 & -2 \end{bmatrix}$$

are both matrices, and we would say that $A$ is $2 \times 3$ and that $B$ is $3 \times 4$.

### Matrix Math and Manipulations

Assume you have two general matrices, $A$ which has shape $n \times m$ and $B$ which has shape $p \times q$. Some shapes are compatible for matrix multiplication, and many are not:

- The **inner dimensions** must match to multiply matrices
  - i.e. you may multiply a $2 \times 3$ and a $3 \times 4$ matrix. You may *not* multiply a $2 \times 3$ and $2 \times 3$ matrix.
- **Order matters**
  - based on the above, you should note that if $A$ is $2 \times 3$ and $B$ is $3 \times 4$, you may find $AB$ but **NOT** $BA$.
 
**Also**: matrix multiplication is **NOT** pairwise multiplication (that's the Hadamard product!). This should be obvious from the restrictions above. If you cannot multiply $2\times3$ and $2\times3$, then multiplication can't be that simple. So how do we do it?

In [3]:
A = np.array([[1, 2, 3],
             [-4, -5, -6]])
B = np.array([[-5, 6, 1, -2],
             [0, 1, 1, 0],
             [8, 6, 4, -2]])


In [6]:
# note the error when we try to multiply pairwise elements (* operator)
# you can't take the Hadamard product of matrices of different dimension
# A * B

In [4]:
# the numpy function that does matrix multiplication
np.matmul(A, B)

array([[ 19,  26,  15,  -8],
       [-28, -65, -33,  20]])

### What happened? How does matrix multiplication work? (Demonstration on the whiteboard)

Notice that the resulting matrix from multiplying $A$ by $B$ **kept the outer dimensions of the two matrices**. I.e. multiplying a $2\times3$ matrix by a $3\times4$ matrix resulted in a single $2\times4$ matrix. This is because matrix multiplication is a result of:

- Each element in the product matrix is the **dot product** of the corresponding **row from the left matrix** and **column from the right matrix**

In [12]:
# first row, first column
np.dot(A[0,:], B[:,0])

19

In [13]:
# first row, second column
np.dot(A[0,:], B[:,1])

26

In [14]:
# second row, first column
np.dot(A[1,:], B[:,0])

-28

### So, if vectors are matrices...

As long as the inner dimensions match, we can multiply matrices. This means we can multiply vectors by matrices and matrices by vectors. In fact:

- Matrix-vector multiplcation ($A\vec{x}$, for matrix $A$ and vector $\vec{x}$) is a linear combination of the **rows** of $A$
- Vector-Matrix multiplcation ($\vec{x}A$, for vector $\vec{x}$ and matrix $A$) is a linear combination of the **columns** of $A$

**Example:**

$$A = \begin{bmatrix} 1 & 2 & 3 \\ -4 & -5 & -6 \end{bmatrix}$$
$$\vec{x} = \begin{bmatrix} 2 \\ 4 \\ -2 \end{bmatrix}$$
$$A\vec{x} = \begin{bmatrix} 1 & 2 & 3 \\ -4 & -5 & -6 \end{bmatrix}  \begin{bmatrix} 2 \\ 4 \\ -2 \end{bmatrix} = \begin{bmatrix} 1(2) + 2(4) + 3(-2) \\ (-4)(2) + (-5)(4) + (-6)(-2) \end{bmatrix} = \begin{bmatrix} 4 \\ -16 \end{bmatrix}$$

In [15]:
x = np.array([2, 4, -2])
np.matmul(A, x)

array([  4, -16])

In [16]:
# remember that order matters; we cannot do xA because x is 3x1 and A is 2x3:
#np.matmul(x, A)

Think for a moment about the structure of our penguin data:

In [17]:
df_penguin_num_scaled.shape

(333, 4)

This is the **dimension** of the matrix that represents all our data; 333 countries with 4 numeric features measured about each one. Since we're going to be treating it as a matrix, it would make sense to cast it to an **array** in Python using NumPy. Note that you can't do this if any of the columns of your data are strings/categorical, because arrays, just like matrices, can only have numbers in them:

In [12]:
Xdat = np.array(df_penguin_num_scaled)
Xdat

array([[-0.89469547,  0.77955895, -1.42460769, -0.56762058],
       [-0.82155152,  0.11940428, -1.06786655, -0.50552542],
       [-0.67526362,  0.42409105, -0.42573251, -1.18857213],
       ...,
       [ 1.1716211 , -0.74387492,  1.50066961,  1.91618562],
       [ 0.22074976, -1.20090508,  0.78718734,  1.23313892],
       [ 1.08019116, -0.5407504 ,  0.85853557,  1.48151954]])

### Matrix-vector Multiplication as a Transformation

Let us focus only on the situation where we have a matrix multiplied by a column vector, as above, but let's simplify (for now) to the situation where the matrix $A$ is a **square** matrix (i.e. same number of rows and columns). This means that the vector $\vec{x}$ must have the same number of rows as $A$, and the resulting product $A\vec{x}$ will be of the same dimension as $\vec{x}$:

**Example:**

$$A = \begin{bmatrix} 2 & 0 \\ 0 & 3 \end{bmatrix}$$
$$\vec{x} = \begin{bmatrix} 2 \\ 4 \end{bmatrix}$$
$$A\vec{x} = \begin{bmatrix} 2 & 0 \\ 0 & 3 \end{bmatrix} \begin{bmatrix} 2 \\ 4 \end{bmatrix} = \begin{bmatrix} 2(2) + 0(4) \\ 0(2) + 3(4) \end{bmatrix} = \begin{bmatrix} 4 \\ 12 \end{bmatrix}$$

In [13]:
A = np.array([[2,0],
              [0,3]])
x = np.array([2, 4])
np.matmul(A, x)

array([ 4, 12])

What happens if we think about this in terms of **how $A$ *changes* $\vec{x}$**? Before applying the matrix $A$ to $\vec{x}$, it was $\vec{x} = \begin{bmatrix} 2 \\ 4 \end{bmatrix}$, and afterwards it was $A\vec{x} = \begin{bmatrix} 4 \\ 12 \end{bmatrix}$, which we note is a vector of the same dimension. You can think of this graphically as $A$ **transforming** $\vec{x}$.

(Draw a picture on the board)

Now what happens if we apply a different matrix, $B$? Perhaps something like:

$$B = \begin{bmatrix} 0 & -1 \\ 1 & 0 \end{bmatrix}$$
$$\vec{x} = \begin{bmatrix} 2 \\ 4 \end{bmatrix}$$
$$B\vec{x} = \begin{bmatrix} 0 & -1 \\ 1 & 0 \end{bmatrix} \begin{bmatrix} 2 \\ 4 \end{bmatrix} = \begin{bmatrix} 0(2) + (-1)(4) \\ 1(2) + 0(4) \end{bmatrix} = \begin{bmatrix} -4 \\ 2 \end{bmatrix}$$

In [20]:
B = np.array([[0,-1],
              [1,0]])
np.matmul(B, x)

array([-4,  2])

(Draw a picture on the board)

The values in the **diagonal** elements of $A$ seem to **scale** the vector $\vec{x}$ when applied, and the values in the **off-diagonal** elements of $B$ seem to **rotate** the vector $\vec{x}$ when applied. So, if we combine them, what happens?

$$A = \begin{bmatrix} 2 & 0 \\ 0 & 3 \end{bmatrix}$$
$$B = \begin{bmatrix} 0 & -1 \\ 1 & 0 \end{bmatrix}$$
$$\vec{x} = \begin{bmatrix} 2 \\ 4 \end{bmatrix}$$
$$AB\vec{x} = ??$$

- Remember that order matters, so the above $AB\vec{x}$ will **first rotate, then scale**, while $BA\vec{x}$ will **first scale, then rotate**. Note that this *does* make a difference in this case:

In [21]:
np.matmul(A, np.matmul(B, x))

array([-8,  6])

In [22]:
np.matmul(B, np.matmul(A, x))

array([-12,   4])

(Draw pictures on the board to illustrate)

Finally note that if you tried to combine the two actions into one matrix, $C$ (by adding the matrices together), the matrix operations are not "additive":

In [23]:
C = np.array([[2, -1],
              [1, 3]])
np.matmul(C, x)

array([ 0, 14])

### Vector Spaces and Spans

![d](https://i.redd.it/3b82i66pulz81.jpg)
What is a vector space?

- The coordinate planes defined by the dimensions make up the vector space; i.e. the number line makes up the 1-dimensional "vector" space, the $x-y$ axes make up the 2-dimensional vector space (a plane), while the $x-y-z$ axes make up the 3-dimensional vector space, etc.

The **basis vectors** of a vector space are the vectors that "define" the direction of the axes, for example:

- the $x-y$ plane has basis vectors: $\hat{i} = \begin{bmatrix} 1 \\ 0 \end{bmatrix}$ and $\hat{j} = \begin{bmatrix} 0 \\ 1 \end{bmatrix}$
  - any other vector in 2-dimensional space can be reached by a linear combination of the two basis vectors.
 
**Example:**

$$\begin{bmatrix} 3 \\ 4 \end{bmatrix} = 3\hat{i} + 4\hat{j}$$
$$\begin{bmatrix} -23 \\ 42 \end{bmatrix} = -23\hat{i} + 42\hat{j}$$

In [24]:
ihat = np.array([1,0])
jhat = np.array([0,1])
3*ihat + 4*jhat

array([3, 4])

In [25]:
-23*ihat + 42*jhat

array([-23,  42])

You can use *other* vectors as "basis" vectors. For example, you can determine which coordinates can be reached by a linear combination of the following two vectors:

$$\vec{v} = \begin{bmatrix} 2 \\ 0 \end{bmatrix}$$
$$\vec{w} = \begin{bmatrix} 1 \\ 2 \end{bmatrix}$$

This is called finding the **span** of a set of vectors. The **span of the basis vectors $\hat{i}$ and $\hat{j}$ is the entire $x-y$ plane**. How do you determine the span of a set of vectors? Use placeholders:

$$\alpha \vec{v} + \beta \vec{w} = \alpha \begin{bmatrix} 2 \\ 0 \end{bmatrix} + \beta \begin{bmatrix} 1 \\ 2 \end{bmatrix} = \begin{bmatrix} 2\alpha + \beta \\ 2\beta \end{bmatrix}$$

This is a 2-dimensional vector where (using $x$ for the $x$-axis and $y$ for the $y$-axis), $x = 2\alpha + \beta$ or $y = 2\beta$. These functions help us **define the span**; note that there are no restrictions on what $x$ and $y$ can be (given any choices of $\alpha$ and $\beta$), meaning that these two vectors' span is also the entire $x-y$ plane:

$$y = 2(x - 2\alpha) \rightarrow y = 2x - 4\alpha$$

**Example (when the span is *not* the entire plane):**

$$\vec{v} = \begin{bmatrix} -1 \\ -2 \end{bmatrix}$$
$$\vec{w} = \begin{bmatrix} 2 \\ 4 \end{bmatrix}$$
$$\alpha \vec{v} + \beta \vec{w} = \alpha \begin{bmatrix} -1 \\ -2 \end{bmatrix} + \beta \begin{bmatrix} 2 \\ 4 \end{bmatrix} = \begin{bmatrix} -\alpha + 2\beta \\ -2\alpha + 4\beta \end{bmatrix}$$

Which means $x = -\alpha + 2\beta$ and $y = -2\alpha + 4\beta$, which in the 2-dimensional vector space is simply the **line $y=2x$**:

$$y = -2\alpha + 4\beta = 2(-\alpha + 2\beta) = 2x$$

**Fact/Note:** if a 2-d vector is a multiple of the other, you are guaranteed to have a line as the span of the two vectors (above, for example, $\vec{w} = -2\vec{v}$).

#### Spans In Summary

In two-dimensions, there are three cases for the span of any set of 2-dimensional vectors:

- Every point in the plane (see above)
- A line passing through the origin (see above)
- The origin (special case: the span of a set of origin vectors)

In N-dimensions, there are *still* three cases for the span of any set of $N$-dimensional vectors:

- Every point in the $N$-dimensional space
- A reduced dimensionality space, passing through the origin (e.g. a plane or a line in 3-dimensions)
- The origin

**Finally**: the span of $N$ vectors is never more than $N$-dimensional space.

- Example: The span of any single vector (of any dimension) is either a line or the origin (if it is the origin):

$$\vec{v} = \begin{bmatrix} 1 \\ 2 \\ 3 \end{bmatrix}$$

Even though $\vec{v}$ exists in 3-dimensional space, the span of this single vector are any points reached by: $\alpha \vec{v} = \begin{bmatrix} \alpha \\ 2\alpha \\ 3\alpha \end{bmatrix}$, which is the line $z = x + y$ in 3-dimensional space.

### Linear Dependence and Independence

A set of vectors is **linearly dependent** if one of the vectors is a linear combination of the others:

- i.e. if the span is a line (see above, and below)

A set of vectors is **linearly independent** if each vector adds a new dimension to the span

- see below for general idea

**Linearly Dependent Vectors**:

The set of vectors: $\vec{a} = \begin{bmatrix} 1 \\ 0 \\ 0 \end{bmatrix}$, $\vec{b} = \begin{bmatrix} 0 \\ 1 \\ 0 \end{bmatrix}$, and $\vec{c} = \begin{bmatrix} 2 \\ 3 \\ 0 \end{bmatrix}$ is linearly dependent because $\vec{c} = 2\vec{a} + 3\vec{b}$.

**Linearly Independent Vectors**:

The set of vectors: $\vec{a} = \begin{bmatrix} \alpha \\ 0 \end{bmatrix}$ and $\vec{b} = \begin{bmatrix} \beta \\ \text{Anything Non-Zero} \end{bmatrix}$ for any $\alpha$ and $\beta$ is linearly independent.

**Fact/Note:** $N+1$ or more vectors of length $N$ are linearly dependent. Example:

The set of vectors: $\vec{a} = \begin{bmatrix} 2 \\ 1 \end{bmatrix}$, $\vec{b} = \begin{bmatrix} -4 \\ 6 \end{bmatrix}$, and $\vec{c} = \begin{bmatrix} 0 \\ 1 \end{bmatrix}$ is linearly dependent because $\vec{b} = -2\vec{a} + 8\vec{c}$.

- This is actually a very important point for machine learning with data. When you have more **features** than **observations** (this is called the [Large p, small n problem](https://www.google.com/search?q=Large+p%2C+small+n+problem&sca_esv=576533920&sxsrf=AM9HkKl_8u1hxRNyO9OTf4YbeWyD68GxSw%3A1698255238798&ei=hlE5ZduOMIKU5OMPj5mxOA&ved=0ahUKEwjb6fvh3ZGCAxUCCnkGHY9MDAcQ4dUDCBA&uact=5&oq=Large+p%2C+small+n+problem&gs_lp=Egxnd3Mtd2l6LXNlcnAiGExhcmdlIHAsIHNtYWxsIG4gcHJvYmxlbTIFEAAYgAQyCBAAGIoFGIYDMggQABiKBRiGAzIIEAAYigUYhgNIqSdQ0AZYuiZwBHgBkAEBmAHsAaABpRWqAQYxNy42LjK4AQPIAQD4AQHCAgoQABhHGNYEGLADwgIEECMYJ8ICBxAuGIoFGCfCAggQABiKBRiRAsICCxAuGIoFGLEDGIMBwgIREC4YgAQYsQMYgwEYxwEY0QPCAgcQIxiKBRgnwgILEC4YgwEYsQMYgATCAgsQABiKBRixAxiDAcICCxAAGIAEGLEDGIMBwgIOEC4YgAQYxwEYrwEYjgXCAggQLhiABBixA8ICDhAuGIAEGLEDGMcBGNEDwgIIEAAYgAQYsQPCAgoQABiABBgUGIcCwgIGEAAYFhge4gMEGAAgQYgGAZAGCA&sclient=gws-wiz-serp)) it means that you are almost certainly going to overfit your data (and, that the features are linearly dependent). We can see a practical example of this when we learn line of best fit shortly.

## Showing In/Dependence using Reduced Row Echelon Form (RREF)
### (Tentative: may skip depending on if we're running behind schedule or not)

One quick and easy way to determine if a set of vectors are independent or not (as long as you have only as many vectors as elements and not more) is to put them into a matrix and **row reduce** that matrix. **Row reduction** involves operating on the matrix in specific/limited ways in order to put into a specific form, Reduced Row Echelon Form (RREF) which consists of:

- Diagonal elements of the matrix are all 1 or 0
- All elements above and below the 1's on the diagonal are 0
- Any rows made up entirely of 0's are on the bottom of the matrix

The operations you can perform on the matrix are limited to:

- Scaling rows (i.e. multiplying rows by some non-zero constant)
- Summing/Subtracting rows (i.e. taking one row and adding/subtracting it from another)
- Swapping rows (self explanatory)

Once you are done putting the matrix into RREF, if you have the identity matrix, the vectors are independent; if there are any 0 rows, the vectors are dependent. 

**Example**: We know that the set of vectors  $\vec{a} = \begin{bmatrix} 1 \\ 0 \\ 0 \end{bmatrix}$, $\vec{b} = \begin{bmatrix} 0 \\ 1 \\ 0 \end{bmatrix}$, and $\vec{c} = \begin{bmatrix} 2 \\ 3 \\ 0 \end{bmatrix}$ is linearly dependent because $\vec{c} = 2\vec{a} + 3\vec{b}$; can we show this by finding the RREF?

$$\left[
    \begin{array}{ccc}
        1 & 0 & 2\\
        0 & 1 & 3\\
        0 & 0 & 0
    \end{array}
\right]$$

Well... that was easy! In fact, the matrix is already **IN** RREF, and there is a zero row. What about a more complicated example? Let's use $\vec{a} = \begin{bmatrix} 1 \\ 2 \\ 0 \end{bmatrix}$, $\vec{b} = \begin{bmatrix} -1 \\ 1 \\ -1 \end{bmatrix}$, and $\vec{c} = \begin{bmatrix} 1 \\ -4 \\ 1 \end{bmatrix}$:

These vectors are independent! This also happens to mean that their span is all of 3-dimensional space (since, you could reach any 3-dimensional vector you'd like to with a linear combination of them). Notice that wasn't true of the first set of three vectors; if a set of vectors is dependent, the span is a lower dimensional space.

$$\left[
    \begin{array}{ccc}
        1 & -1 & 1 \\
        2 & 1 & -4 \\
        0 & -1 & 1 
    \end{array}
\right]
        \xrightarrow[]{r_1'= (r_1 - 2r_0)/3}
        \left[
    \begin{array}{ccc}
        1 & -1 & 1 \\
        0 & 1 & -2 \\
        0 & -1 & 1 
    \end{array}
\right]
        \xrightarrow[r_2'=-(r_2+r_1)]{r_0'=r_0 + r_1}
        \left[
    \begin{array}{ccc}
        1 & 0 & -1 \\
        0 & 1 & -2 \\
        0 & 0 & 1 
    \end{array}
\right]
        \xrightarrow[r_1'=r_1 + 2 r_2]{r_0': r_0 + r_2}
        \left[
    \begin{array}{ccc}
        1 & 0 & 0 \\
        0 & 1 & 0 \\
        0 & 0 & 1 
    \end{array}
\right]$$

We can avoid doing the RREF by hand, and instead use the Python module `sympy` which does it very quickly for us by casting our NumPy array to a `Matrix` object from that package and using the `.rref` function:

In [26]:
!pip install sympy

In [27]:
from sympy import *

M = Matrix(np.array([[1, -1, 1], [2, 1, -4], [0, -1, 1]]))
# Use sympy.rref() method  
M_rref = M.rref()   
      
M_rref[0]

Matrix([
[1, 0, 0],
[0, 1, 0],
[0, 0, 1]])

This helps us figure out very quickly if much larger sets of vectors are independent. For example, what about these four 4-dimensional vectors:

$\vec{a} = \begin{bmatrix} -1 \\ 2 \\ 1 \\ -2 \end{bmatrix}$, $\vec{b} = \begin{bmatrix} 0 \\ 1 \\ 0 \\ 1 \end{bmatrix}$, $\vec{c} = \begin{bmatrix} -2 \\ 1 \\ 3 \\ 4 \end{bmatrix}$, and $\vec{d} = \begin{bmatrix} -3 \\ 3 \\ 4 \\ 2 \end{bmatrix}$

In [28]:
M = Matrix(np.array([[-1, 0, -2, -3], [2, 1, 1, 3], [1, 0, 3, 4], [-2, 1, 4, 2]]))
M

Matrix([
[-1, 0, -2, -3],
[ 2, 1,  1,  3],
[ 1, 0,  3,  4],
[-2, 1,  4,  2]])

In [29]:
# Use sympy.rref() method  
M_rref = M.rref()   
      
M_rref[0]

Matrix([
[1, 0, 0, 1],
[0, 1, 0, 0],
[0, 0, 1, 1],
[0, 0, 0, 0]])

### Eigenvalues and Eigenvectors

An **eigenvector** is a vector that for a given transformation, doesn't move off its own span, or in a simple case:

- a non-zero vector $\vec{v}$ associated with a square matrix $A$ such that

$$A\vec{v} = \lambda\vec{v}$$

In other words, multiplying $\vec{v}$ by $A$ only scales $\vec{v}$ by $\lambda$.

- $\lambda$ is called the **eigenvalue** of the **eigenvector** $\vec{v}$

How would we use this?

**Example**

Find the eigenvalues and eigenvectors for the given matrix:

$$A = \begin{bmatrix}
    -1 & 3 \\ 0 & 2
\end{bmatrix}$$

We know:

$$A\vec{v} = \lambda\vec{v}$$

And we wish to find both $\lambda$ and $\vec{v}$. We can rewrite:

$$A\vec{v} - \lambda\vec{v} = \vec{0}$$
$$(A - \lambda I)\vec{v} = \vec{0}$$

How do we determine what the values of $\lambda$ are? Notice that we want to, essentially, reduce the space of the vector $\vec{v}$ to the origin. **We want the matrix it is multiplied by to have a determinant of zero**:

$$det(A - \lambda I) = 0$$
$$A - \lambda I = \begin{bmatrix}
    -1-\lambda & 3 \\ 0 & 2-\lambda
\end{bmatrix}$$
$$det(A - \lambda I) = (-1-\lambda)(2-\lambda) - (3)(0) = 0$$
$$(-1-\lambda)(2-\lambda) = 0$$

This implies that the eigenvalues are $\lambda = \{-1,2\}$.

And the eigenvectors? Use each of the eigenvalues to figure them out:

$$\lambda = -1$$
$$\begin{bmatrix}
    -1-(-1) & 3 \\ 0 & 2-1
\end{bmatrix}\begin{bmatrix}
    x \\ y
\end{bmatrix} = \begin{bmatrix}
    0 \\ 0
\end{bmatrix}$$
$$\begin{bmatrix}
    0 \\ 0
\end{bmatrix}x + \begin{bmatrix}
    3 \\ 3
\end{bmatrix}y = \begin{bmatrix}
    0\\0
\end{bmatrix}$$

When $\lambda=-1$, the eigenvector(s) are anything on the $x$-axis.

$$\lambda = 2$$
$$\begin{bmatrix}
    -1-(2) & 3 \\ 0 & 2-2
\end{bmatrix}\begin{bmatrix}
    x \\ y
\end{bmatrix} = \begin{bmatrix}
    0 \\ 0
\end{bmatrix}$$
$$\begin{bmatrix}
    -3 \\ 0
\end{bmatrix}x + \begin{bmatrix}
    3 \\ 0
\end{bmatrix}y = \begin{bmatrix}
    0\\0
\end{bmatrix}$$
$$-3x + 3y = 0$$
$$-x+y=0$$
$$x=y$$

When $\lambda = 2$, the eigenvector(s) are all scaled versions of $\begin{bmatrix}
    1\\1
\end{bmatrix}$. For simplicity's sake, we usually just say $\begin{bmatrix}
    1\\1
\end{bmatrix}$, since:


 - **Any scaled version of an eigenvector is also an eigenvector with the same eigenvalue**

In [30]:
# we can use NumPy to find eigenvalue and eigenvector pairs easily
A = np.array([[-1, 3], [0, 2]])

print('eigenvalues:', np.linalg.eig(A)[0])

eigenvalues: [-1.  2.]


In [31]:
# format the eigenvector matrix; the columns are the eigenvectors corresponding to the eigenvalues
# note that these any multiple of each column is also an eigenvector
Matrix(np.linalg.eig(A)[1])

Matrix([
[1.0, 0.707106781186547],
[  0, 0.707106781186547]])

## Lecture Break/Practice

Find the eigenvalues and corresponding eigenvectors of the matrix:

$$A = \begin{bmatrix}
    1 & 5 \\ 2 & 4
\end{bmatrix}$$

Do it first by hand, then use NumPy to verify.

### Orthogonality

Vectors are **orthogonal** if their dot product is zero (equivalently, if the angle between them is 90 degrees).

**Examples:**

In [32]:
# the basis vectors are orthogonal
np.dot(ihat, jhat)

0

In [33]:
northwest = np.array([-1,1])
northeast = np.array([1, 1])
np.dot(northwest, northeast)

0

## Next Time: Projections (or, where we start finally approaching Machine Learning)